### DISCUSSION TITLE
does `latest!-1` make sense?

### DISCUSSION THREAD

**Discussion by jdx** [jdx/mise#340] [General]
I added the ability to subtract versions from aliases which is helpful to express things like "I want the latest 2 versions of nodejs", for example. Here is what it looks like currently:

```
# use lts version of nodejs
nodejs lts

# install 2 versions behind the latest lts
# (e.g.: 16 if lts is 18)
nodejs lts!-2

# install python-3.10 if the latest is 3.11
python latest!-0.1
```

What do people think? This is experimental so it can change. One thought I have is this isn't compatible with the other types of `.tool-versions` syntaxes like the following:

```
erlang  ref:master   # compile from vcs ref
golang  prefix:1.19  # uses the latest 1.19.x version—needed in case "1.19" is an exact match
shfmt   path:./shfmt # use a custom runtime
```

So maybe it would be better to do something like `minus-major:lts` or `sub-1:lts`? That might fit a bit better with existing alternatives.

**Comment by JoiHallecksM6** [comment]
The ability to subtract versions from aliases is a useful addition to the syntax and can make it easier for users to specify which versions they want to use. However, as you have noted, it may not be fully compatible with other syntaxes that are already in use.

Using "minus-major:lts" or "sub-1:lts" instead of "lts!-2" may indeed make the syntax more compatible with existing alternatives. It may also be helpful to consider adding documentation and examples for how to use this new syntax to avoid confusion for users who are not familiar with it.

Overall, it's great to see experimentation and innovation in syntax to make it more user-friendly and flexible. As long as the changes are well-documented and communicated clearly, they can provide significant benefits to users.

**Reply by jdx** [reply]
`minus-major:ts` doesn't work since you need to be able subtract 2 in the case of even/odd languages like nodejs. However `sub-1:lts` I think I like a lot more with the benefit of hindsight than the current syntax since it fits with the other types like [`path:`, `prefix:`, and `ref:`](https://github.com/jdxcode/rtx/blob/24f239b16252ce48c0e8bca614705375f1c71deb/src/config/config_file/tool_versions.rs#L126-L129). That would clean up the internal code a fair bit and also more consistent which would help people understand how to use this stuff.

Currently this is documented in the [`.tool-versions`](https://github.com/jdxcode/rtx#tool-versions) section, which made sense initially since that was the main place this was defined, however now with `.rtx.toml` it might be better if I called it out explicitly. The main reason I haven't done that is I just haven't figured out what to call these (as a collection) yet. "version types?" I feel that is too generic. In the code I call them "ToolVersionType".

**Reply by jdx** [reply]
started working on this: #556

**Comment by Marukome0743** [comment]
The syntax discussed here was eventually changed in [PR #556](https://github.com/jdx/mise/pull/556). Current mise uses the more explicit `sub-<version>:<original>` form instead of `latest!-1`.

Examples:

```text
node    sub-2:lts
python  sub-0.1:latest
```

`sub-2:lts` resolves the `lts` alias and subtracts 2 from its major component. `sub-0.1:latest` subtracts 1 from the resolved minor component.

Multiple versions can be configured together:

```toml
[tools]
node = ["lts", "sub-2:lts"]
python = ["latest", "sub-0.1:latest"]
```

I also checked the current main branch with:

```sh
mise ls-remote tiny@sub-1:3
```

It correctly resolved the `3` selector to the matching `2.x` versions.

This feature performs numeric component subtraction after resolving the original selector. It should therefore only be used for tools whose version scheme supports that operation; mise does not generally assume that every tool uses semantic versions.

The current syntax is documented here:

https://mise.en.dev/configuration.html#tool-versions

*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*

**Comment by JamBalaya56562** [comment]
Adding one thing to @Marukome0743's answer, because the two examples in it do not work on the current release.

Measured on **v2026.8.2**:

```console
$ mise ls-remote node@sub-2:lts
The application panicked (crashed).
Message:  called `Option::unwrap()` on a `None` value

$ mise ls-remote python@sub-0.1:latest
The application panicked (crashed).
Message:  called `Option::unwrap()` on a `None` value
```

A numeric base is fine — `mise ls-remote node@sub-1:22` returns `21.7.3`, which matches the `tiny@sub-1:3` check above. It is specifically an **alias or `latest` as the base** that crashes, because the subtraction ran against the unresolved string instead of a version.

#11796 fixes it: the base is resolved first (through aliases, and through `latest` to a concrete version), and a base that still is not numeric now produces an error naming the version rather than a panic. Merged, but **not yet in a release** — v2026.8.3 was cut before it landed.
